blues/sparrow-reference-web-app
# feat: date range selector for charts

# Problem Context

After talking with the design team, this story is pivoting away from zoomable charts (deemed a little small to be useful in their current state and possibly better suited to a "chart details" view where a user can really dig into the specifics of one chart in particular), and is now looking at the possibilities for a range picker to adjust the chart data displayed on the sensor details page.

Thinks a select option where users can choose from: Last Hour, Last 12 Hours, Last day, Last 7 days, Last month, Last quarter, etc. and the charts will adjust to show that time frame based on what is selected from the dropdown.

## Changes

Add a new select option to the Node details page to allow users to select from different pre-determined timeframes to view their chart data. Made `startDate` an option configurable in the UI that gets stored as part of the URLs query string to force a refresh on the backend with the added date range parameter. Also added an Info icon to the motion count reading at the top of the page to display the total motion tooltip on hover that was previously added.
Updated the env variable `HUB_HISTORICAL_DATA_START_DATE` to be a string and to be comprised of minutes (instead of days) to allow for easier swapping between server side env vars and client side query params and formatting of the date range start from minutes to epoch time.
Add a few new unit tests for forms and existing test suites to account for new functionality.
Rename some straggler "Sensor"-related files to be "Node"-files (chart components, sensor schema, sensor reading schema, and sensor reading schemas)

## Screenshot (if applicable)

![Screen Shot 2022-03-09 at 1 59 11 PM](https://user-images.githubusercontent.com/20400845/157512588-6a154ba2-78b4-4602-a321-9aa51dc0d3aa.png)

## Testing

Unit / integration / e2e tests?

A few new unit tests added, and e2es updated - attempting to make them less flaky as well, but in GH Actions they're still iffy.

![Screen Shot 2022-03-09 at 11 28 25 AM](https://user-images.githubusercontent.com/20400845/157512161-df8faac8-04c3-44b8-aae9-1921b21562cf.png)

Steps to test manually?

1. Go to http://localhost:4000
2. Navigate to a node that has charts of data.
3. Use the new dropdown to select chart data for various time frames.
4. Watch the charts update to reflect this new amount of data and see the query stored in the URL string

Any other sorts of testing notes to include?

NOTE: Please update your `.env.local` file's to read `HUB_HISTORICAL_DATA_START_DATE="1440"` - the data is now being translated to minutes instead of days for better granularity when displaying chart values, so 1440 minutes = 1 day.

## Any other related PRs

n/a

## Ticket(s)

https://www.pivotaltracker.com/story/show/181500973
